File: tests/rosetta/transpiler/php/cartesian-product-of-two-or-more-lists-2.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function listStr($xs) {
  $s = '[';
  $i = 0;
  while ($i < count($xs)) {
  $s = $s . _str($xs[$i]);
  if ($i < count($xs) - 1) {
  $s = $s . ' ';
}
  $i = $i + 1;
};
  $s = $s . ']';
  return $s;
};
  function llStr($lst) {
  $s = '[';
  $i = 0;
  while ($i < count($lst)) {
  $s = $s . listStr($lst[$i]);
  if ($i < count($lst) - 1) {
  $s = $s . ' ';
}
  $i = $i + 1;
};
  $s = $s . ']';
  return $s;
};
  function cartN($lists) {
  if ($lists == null) {
  return [];
}
  $a = $lists;
  if (count($a) == 0) {
  return [[]];
}
  $c = 1;
  foreach ($a as $xs) {
  $c = $c * count($xs);
};
  if ($c == 0) {
  return [];
}
  $res = [];
  $idx = [];
  foreach ($a as $_) {
  $idx = array_merge($idx, [0]);
};
  $n = count($a);
  $count = 0;
  while ($count < $c) {
  $row = [];
  $j = 0;
  while ($j < $n) {
  $row = array_merge($row, [$a[$j][$idx[$j]]]);
  $j = $j + 1;
};
  $res = array_merge($res, [$row]);
  $k = $n - 1;
  while ($k >= 0) {
  $idx[$k] = $idx[$k] + 1;
  if ($idx[$k] < count($a[$k])) {
  break;
}
  $idx[$k] = 0;
  $k = $k - 1;
};
  $count = $count + 1;
};
  return $res;
};
  function main() {
  echo rtrim(llStr(cartN([[1, 2], [3, 4]]))), PHP_EOL;
  echo rtrim(llStr(cartN([[3, 4], [1, 2]]))), PHP_EOL;
  echo rtrim(llStr(cartN([[1, 2], []]))), PHP_EOL;
  echo rtrim(llStr(cartN([[], [1, 2]]))), PHP_EOL;
  echo rtrim(''), PHP_EOL;
  echo rtrim('['), PHP_EOL;
  foreach (cartN([[1776, 1789], [7, 12], [4, 14, 23], [0, 1]]) as $p) {
  echo rtrim(' ' . listStr($p)), PHP_EOL;
};
  echo rtrim(']'), PHP_EOL;
  echo rtrim(llStr(cartN([[1, 2, 3], [30], [500, 100]]))), PHP_EOL;
  echo rtrim(llStr(cartN([[1, 2, 3], [], [500, 100]]))), PHP_EOL;
  echo rtrim(''), PHP_EOL;
  echo rtrim(llStr(cartN(null))), PHP_EOL;
  echo rtrim(llStr(cartN([]))), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = intdiv($__end - $__start, 1000);
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;;

File: tests/rosetta/transpiler/FS/cartesian-product-of-two-or-more-lists-2.fs
// Generated 2025-07-27 22:23 +0700

exception Break
exception Continue

exception Return

let rec listStr (xs: int array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable xs = xs
    try
        let mutable s: string = "["
        let mutable i: int = 0
        while i < (unbox<int> (Array.length xs)) do
            s <- s + (string (xs.[i]))
            if i < (unbox<int> ((unbox<int> (Array.length xs)) - 1)) then
                s <- s + " "
            i <- i + 1
        s <- s + "]"
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
and llStr (lst: int array array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable lst = lst
    try
        let mutable s: string = "["
        let mutable i: int = 0
        while i < (unbox<int> (Array.length lst)) do
            s <- s + (unbox<string> (listStr (unbox<int array> (lst.[i]))))
            if i < (unbox<int> ((unbox<int> (Array.length lst)) - 1)) then
                s <- s + " "
            i <- i + 1
        s <- s + "]"
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
and cartN (lists: obj) =
    let mutable __ret : int array array = Unchecked.defaultof<int array array>
    let mutable lists = lists
    try
        if lists = null then
            __ret <- unbox<int array array> [||]
            raise Return
        let a: int array array = unbox<int array array> lists
        if (unbox<int> (Array.length a)) = 0 then
            __ret <- unbox<int array array> [|[||]|]
            raise Return
        let mutable c: int = 1
        for xs in a do
            c <- c * (Seq.length xs)
        if c = 0 then
            __ret <- unbox<int array array> [||]
            raise Return
        let mutable res: int array array = [||]
        let mutable idx: int array = [||]
        for _ in a do
            idx <- unbox<int array> (Array.append idx [|0|])
        let mutable n: int = Array.length a
        let mutable count: int = 0
        try
            while count < c do
                let mutable row: int array = [||]
                let mutable j: int = 0
                while j < n do
                    row <- unbox<int array> (Array.append row [|(a.[j]).[idx.[j]]|])
                    j <- j + 1
                res <- unbox<int array array> (Array.append res [|row|])
                let mutable k: int = n - 1
                try
                    while k >= 0 do
                        idx.[k] <- (unbox<int> (idx.[k])) + 1
                        if (unbox<int> (idx.[k])) < (Seq.length (a.[k])) then
                            raise Break
                        idx.[k] <- 0
                        k <- k - 1
                with
                | Break -> ()
                | Continue -> ()
                count <- count + 1
        with
        | Break -> ()
        | Continue -> ()
        __ret <- unbox<int array array> res
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        printfn "%s" (llStr (cartN [|[|1; 2|]; [|3; 4|]|]))
        printfn "%s" (llStr (cartN [|[|3; 4|]; [|1; 2|]|]))
        printfn "%s" (llStr (cartN [|[|1; 2|]; [||]|]))
        printfn "%s" (llStr (cartN [|[||]; [|1; 2|]|]))
        printfn "%s" ""
        printfn "%s" "["
        for p in cartN [|[|1776; 1789|]; [|7; 12|]; [|4; 14; 23|]; [|0; 1|]|] do
            printfn "%s" (" " + (unbox<string> (listStr (unbox<int array> p))))
        printfn "%s" "]"
        printfn "%s" (llStr (cartN [|[|1; 2; 3|]; [|30|]; [|500; 100|]|]))
        printfn "%s" (llStr (cartN [|[|1; 2; 3|]; [||]; [|500; 100|]|]))
        printfn "%s" ""
        printfn "%s" (llStr (cartN null))
        printfn "%s" (llStr (cartN [||]))
        __ret
    with
        | Return -> __ret
main()
